File: bug34_early_loop_return/buggy/main.js
function compute() {
    let seeds = [0xDEADn, 0xBEEFn, 0xCAFEn, 0xBABEn, 0xFACEn,
        0xFEEDn, 0xD00Dn, 0xC0DEn, 0xACE2n, 0xBEADn];
    let acc = 0n;
    for (let i = 0n; i < 10n; i++) {
        acc ^= seeds[Number(i)] * (i + 1n);
        acc = BigInt.asUintN(64, acc * 0x9E3779B97F4A7C15n);
        let shiftLeft = BigInt(i % 7n + 1n);
        let shiftRight = BigInt(63n - i % 7n);
        acc = BigInt.asUintN(64, (acc << shiftLeft) | (acc >> shiftRight));
        return acc;
    }
    return acc;
}

function correctCompute() {
    let seeds = [0xDEADn, 0xBEEFn, 0xCAFEn, 0xBABEn, 0xFACEn,
        0xFEEDn, 0xD00Dn, 0xC0DEn, 0xACE2n, 0xBEADn];
    let acc = 0n;
    for (let i = 0n; i < 10n; i++) {
        acc ^= seeds[Number(i)] * (i + 1n);
        acc = BigInt.asUintN(64, acc * 0x9E3779B97F4A7C15n);
        let shiftLeft = BigInt(i % 7n + 1n);
        let shiftRight = BigInt(63n - i % 7n);
        acc = BigInt.asUintN(64, (acc << shiftLeft) | (acc >> shiftRight));
    }
    return acc;
}

function run() {
    let result = compute();
    let correct = correctCompute();
    if (result === correct) {
        throw new Error("AssertionError: early return gives same result: undetectable");
    }
    throw new Error(`ValueError: returned inside loop too early: got 0x${result.toString(16)}, expected 0x${correct.toString(16)}`);
}
run();

File: bug34_early_loop_return/buggy/main.java
public class main {
    public static long compute() {
        long[] seeds = { 0xDEADL, 0xBEEFL, 0xCAFEL, 0xBABEL, 0xFACEL,
                0xFEEDL, 0xD00DL, 0xC0DEL, 0xACE2L, 0xBEADL };
        long acc = 0;
        for (int i = 0; i < seeds.length; i++) {
            acc ^= seeds[i] * (i + 1);
            acc *= 0x9E3779B97F4A7C15L;
            acc = (acc << (i % 7 + 1)) | (acc >>> (63 - i % 7));
            return acc;
        }
        return acc;
    }

    public static long correctCompute() {
        long[] seeds = { 0xDEADL, 0xBEEFL, 0xCAFEL, 0xBABEL, 0xFACEL,
                0xFEEDL, 0xD00DL, 0xC0DEL, 0xACE2L, 0xBEADL };
        long acc = 0;
        for (int i = 0; i < seeds.length; i++) {
            acc ^= seeds[i] * (i + 1);
            acc *= 0x9E3779B97F4A7C15L;
            acc = (acc << (i % 7 + 1)) | (acc >>> (63 - i % 7));
        }
        return acc;
    }

    public static void main(String[] args) {
        long result = compute();
        long correct = correctCompute();
        if (result == correct) {
            throw new RuntimeException("AssertionError: early return gives same result: undetectable");
        }
        throw new IllegalArgumentException("ValueError: returned inside loop too early: got 0x"
                + Long.toHexString(result) + ", expected 0x" + Long.toHexString(correct));
    }
}
